force-app/main/default/classes/RD2_SustainerEvaluationService.cls

```diff
@@ -0,0 +1,102 @@
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2022
+* @group Recurring Donations
+* @description Evaluate Sustainer
+*/
+public inherited sharing class RD2_SustainerEvaluationService {
+    private static final String SUSTAINER_FIELD = UTIL_Namespace.StrTokenNSPrefix('Sustainer__c');
+
+    @TestVisible
+    private static RD2_QueryService queryService {
+        get {
+            if (queryService == null) {
+                queryService = new RD2_QueryService();
+            }
+            return queryService;
+        }
+        set;
+    }
+
+    public enum SustainerType {
+        Active,
+        Lapsed,
+        Former
+    }
+
+    public List<SObject> evaulateSustainersFor(String sObjectType, Set<Id> recordIds) {
+        List<SObject> evaulatedRecords = queryService.getSObjectWithActiveRecurringDonations(sObjectType,  recordIds);
+        List<SObject> toUpdate = new List<SObject>();
+        
+        List<SObject> noneActiveSustainers = calculateSustainerFor(SustainerType.Active, evaulatedRecords, toUpdate);
+
+        if (noneActiveSustainers.size() == 0) {
+            return toUpdate;
+        }
+
+        List<SObject> lapsedSustainerCandidates = queryService.getSObjectWithLapsedRecurringDonations(sObjectType, UTIL_SObject.extractIds(noneActiveSustainers));
+        List<SObject> formerSustainers = calculateSustainerFor(SustainerType.Lapsed, lapsedSustainerCandidates, toUpdate);
+
+        for (SObject formerSustainer : formerSustainers) {
```

Comment:
Only RD-related Contacts/Accounts will execute this code. Therefore, every Contact/Account which is not an 'Active' or 'Lapsed' Sustainer will be a 'Former' Sustainer. The Contact Merge WI will handle None Sustainer logic.